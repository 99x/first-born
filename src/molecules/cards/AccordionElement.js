import React, { Component } from "react";
import {
    View,
    StyleSheet,
    TouchableOpacity,
    Animated,
    Platform
} from "react-native";
import PropTypes from "prop-types";

import { Text } from "../../atoms/Text";
import { Icon } from "../../atoms/Icon";
import { commonColors } from "../../utils/color";

export class AccordionElement extends Component {
    constructor(props) {
        super(props);
        this.state = {
            expanded: props.expanded
        };
        this.fadeAnim = new Animated.Value(0);
    }

    animate() {
        this.fadeAnim = new Animated.Value(0);
        Animated.timing(this.fadeAnim, {
            toValue: 1,
            duration: 250,
            useNativeDriver: true
        }).start();
    }

    handleShowHide = () => {
        this.setState({ expanded: !this.state.expanded });
        this.animate();
    };

    render() {
        const {
            title,
            description,
            image,
            onPress,
            block,
            bodyColor,
            headerColor,
            style,
            isList,
            ...otherProps
        } = this.props;

        const { expanded } = this.state;

        let cardStyle = [
            styles.container,
            { borderColor: headerColor, borderWidth: 1 }
        ];
        const headerStyle = [
            styles.titleContainer,
            { backgroundColor: headerColor }
        ];
        const bodyStyle = {
            ...styles.bodyContainer,
            backgroundColor: bodyColor
        };

        if (!isList) {
            cardStyle.push({ borderRadius: 5 });
        }

        if (style) {
            cardStyle.push(style);
        }

        return (
            <View style={cardStyle}>
                <TouchableOpacity
                    onPress={this.handleShowHide}
                    style={headerStyle}
                    {...otherProps}
                >
                    <Text
                        color={commonColors.darkGrey}
                        bold
                        style={{ flex: 1 }}
                    >
                        {title}
                    </Text>
                    {this.renderIcon()}
                </TouchableOpacity>
                {expanded && (
                    <Animated.View
                        style={[bodyStyle, { opacity: this.fadeAnim }]}
                    >
                        <Text size="footnote" color={commonColors.darkGrey}>
                            {description}
                        </Text>
                    </Animated.View>
                )}
            </View>
        );
    }

    renderIcon() {
        const { expandedIcon, icon } = this.props;
        const { expanded } = this.state;
        return (
            <Icon
                name={expanded ? expandedIcon : icon}
                size={20}
                color={commonColors.inputGrey}
            />
        );
    }
}

AccordionElement.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    bodyColor: PropTypes.string,
    headerColor: PropTypes.string,
    expanded: PropTypes.bool,
    expandedIcon: PropTypes.string,
    icon: PropTypes.string,
    isList: PropTypes.bool,
    ...TouchableOpacity.propTypes
};

AccordionElement.defaultProps = {
    bodyColor: commonColors.white,
    headerColor: commonColors.lightGrey,
    isList: false,
    expanded: false,
    expandedIcon: Platform.OS === "ios" ? "arrow-up" : "arrow-dropup",
    icon: Platform.OS === "ios" ? "arrow-down" : "arrow-dropdown"
};

const styles = StyleSheet.create({
    container: {
        width: "100%",
        flex: 1,
        flexDirection: "column",
        overflow: "hidden"
    },
    titleContainer: {
        paddingVertical: 5,
        paddingHorizontal: 10,
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "center",
        overflow: "hidden"
    },
    bodyContainer: {
        flexDirection: "row",
        flex: 1,
        padding: 20,
        overflow: "hidden"
    }
});
