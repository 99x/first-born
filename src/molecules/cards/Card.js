import React, { Component } from "react";
import { View, StyleSheet, TouchableOpacity, Image, Platform, Dimensions } from "react-native";
import PropTypes from "prop-types";

import { Text } from "../../atoms/Text";
import { commonColors } from "react-native-atom-elements/src/utils/color";

const { width } = Dimensions.get("window");

export class Card extends Component {

    render() {
        const { title, description, image, onPress, block, backgroundColor, style, ...otherProps } = this.props;

        const cardBackColor = backgroundColor ? backgroundColor : commonColors.white;

        let cardStyle = block ? [styles.containerBlock, { backgroundColor: cardBackColor }] : [styles.container, , { backgroundColor: cardBackColor }];

        if (style) {
            cardStyle.push(style);
        }

        return (
            <TouchableOpacity style={cardStyle} disabled={!onPress} onPress={onPress} {...otherProps}>
                {image && <Image source={image} style={styles.image} />}
                <View style={styles.textContainer}>
                    <Text size="h6" color="rgba(33, 33, 33, 0.87)" bold >{title}</Text>
                    {description && <Text size="sub_heading" color="rgba(33, 33, 33, 0.4)">{description}</Text>}
                </View>
            </TouchableOpacity>
        )
    }
}

Card.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    image: PropTypes.any,
    block: PropTypes.bool,
    backgroundColor: PropTypes.string,
    ...TouchableOpacity.propTypes
}

const styles = StyleSheet.create({
    container: {
        width: "100%",
        shadowColor: "#000",
        shadowOffset: { width: 3, height: 3 },
        shadowOpacity: 0.3,
        shadowRadius: 3,
        elevation: 3,
        borderRadius: 3,
        overflow: "hidden",
        marginBottom: 5
    },
    containerBlock: {
        width,
        overflow: "hidden",
        borderTopWidth: 1,
        borderBottomWidth: 1,
        borderTopColor: "rgba(33, 33, 33,  0.4)",
        borderBottomColor: "rgba(33, 33, 33,  0.4)",
        marginBottom: 5
    },
    textContainer: {
        padding: 20,
        flexDirection: 'column',
        justifyContent: 'flex-start',
    },
    image: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        height: 150,
        borderTopLeftRadius: 3,
        borderTopRightRadius: 3
    }
})