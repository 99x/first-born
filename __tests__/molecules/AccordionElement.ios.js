import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { AccordionElement } from "../../src/molecules/cards/AccordionElement";

jest.mock("Platform", () => {
    const Platform = require.requireActual("Platform");
    Platform.OS = "ios";
    return Platform;
});

it("renders accordion element with heading and title", () => {
    const tree = renderer
        .create(<AccordionElement title="Heading" description="description" />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders colored accordion element with heading and title", () => {
    const tree = renderer
        .create(
            <AccordionElement
                title="Heading"
                description="description"
                headerColor="aquamarine"
                bodyColor="aliceblue"
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders accordion element with heading, title and custom expand/collapse icons", () => {
    const tree = renderer
        .create(
            <AccordionElement
                title="Heading"
                description="description"
                icon="add"
                expandedIcon="remove"
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders colored accordion element with heading, title and custom expand/collapse icons", () => {
    const tree = renderer
        .create(
            <AccordionElement
                title="Heading"
                description="description"
                headerColor="aquamarine"
                bodyColor="aliceblue"
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
