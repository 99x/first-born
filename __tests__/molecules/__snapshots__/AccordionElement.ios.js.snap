// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders accordion element with heading and title 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "column",
        "overflow": "hidden",
        "width": "100%",
      },
      Object {
        "borderColor": "rgba(33, 33, 33, 0.4)",
        "borderWidth": 1,
      },
      Object {
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(33, 33, 33, 0.4)",
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
        "overflow": "hidden",
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.5)",
            "fontSize": 20,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders accordion element with heading, title and custom expand/collapse icons 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "column",
        "overflow": "hidden",
        "width": "100%",
      },
      Object {
        "borderColor": "rgba(33, 33, 33, 0.4)",
        "borderWidth": 1,
      },
      Object {
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(33, 33, 33, 0.4)",
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
        "overflow": "hidden",
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.5)",
            "fontSize": 20,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders colored accordion element with heading and title 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "column",
        "overflow": "hidden",
        "width": "100%",
      },
      Object {
        "borderColor": "aquamarine",
        "borderWidth": 1,
      },
      Object {
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "aquamarine",
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
        "overflow": "hidden",
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.5)",
            "fontSize": 20,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders colored accordion element with heading, title and custom expand/collapse icons 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "column",
        "overflow": "hidden",
        "width": "100%",
      },
      Object {
        "borderColor": "aquamarine",
        "borderWidth": 1,
      },
      Object {
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "aquamarine",
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
        "overflow": "hidden",
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.5)",
            "fontSize": 20,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
